### PR TITLE
TypeDoc

### PR DESCRIPTION
**Pull request by henrycatalinismith**
Following up from the Slack discussion. This seems to have reached a point where it's going to move forward better as a pull request instead of just a branch. Curious about the time-related comments! Was tempted to just walk that commit back but it's worth keeping just cos I'm keen to learn more context about that!

### CODE REVIEW

File: src/core/caching/futures.ts [hunk 1]
```diff
@@ -1,8 +1,40 @@
...
+ *   A[data: null\nerror: null\nisLoading: false]
+ *   B[data: null\nerror: null\nisLoading: true]
+ *   C[data: null\nerror: Error\nisLoading: false]
+ *   D[data: Object\nerror: null\nisLoading: false]
+ *   E{Success?}
+ *   A -->|Asynchronous operation begins| B
+ *   B -->|Asynchronous operation ends| E
+ *   E -->|Yes| D
+ *   E -->|No| C
+ * ```
```
Comment (richardolsson) [inline src/core/caching/futures.ts:18]: I'm guessing this should render a flowchart as some sort of graphic, but I'm not sure how to run the code in a way that does. I tried `yarn typedoc` which did build an output into `./typedoc/` that looked like something that I should be able to serve as static files, but that didn't work. What is the right way to view this?

Comment (richardolsson) [reply src/core/caching/futures.ts:18]: An update. It seems that it was just really slow to load (note the loading duration in screenshot below), and one file failed:

image

After more than a minute of loading mermaid-related JS files, it finally rendered this:

image

Is this something that you are also experiencing or is it just me?

**Review by richardolsson — COMMENTED** [review]
This is cool! I have some comments below, and then a general comment which is that I think it would be cool to have this documentation and the Storybook documentation hosted together, perhaps even by the main NEXT.js server, so that it can all be available on a single server (without having to run three, or even two of them).

I wonder if that could be possible by writing a custom build/start script (that we can use on Vercel) that builds Storybook and TypeDoc into the NEXT.js /public directory and then builds and starts NEXT.js. What do you think?

**Comment by richardolsson** [comment]
I did a tiny experiment with my idea and it worked fine!

image

Here's what I did. I just tried building and running it locally, but it should also work fine on Vercel (where we run our development previews) by just changing the build command to `yarn docs:build && yarn build` or whatever.

```diff
diff --git a/.gitignore b/.gitignore
index 4cb8d8bb6..8d2b553c0 100644
--- a/.gitignore
+++ b/.gitignore
@@ -46,4 +46,8 @@ yarn-error.log*
 tsconfig.tsbuildinfo
 
 # typedoc
-/typedoc/
\ No newline at end of file
+/typedoc/
+
+# generated docs
+public/docs
+public/storybook
diff --git a/next.config.js b/next.config.js
index 865a00036..e391413ff 100644
--- a/next.config.js
+++ b/next.config.js
@@ -12,6 +12,16 @@ module.exports = {
   },
   async redirects() {
     return [
+      {
+        source: '/storybook',
+        destination: '/storybook/index.html',
+        permanent: true,
+      },
+      {
+        source: '/docs',
+        destination: '/docs/index.html',
+        permanent: true,
+      },
       {
         source: '/:prevPath*/calendar/events',
         destination: '/:prevPath*/calendar',
diff --git a/package.json b/package.json
index 964649193..61fe05a1d 100644
--- a/package.json
+++ b/package.json
@@ -3,6 +3,7 @@
   "version": "41.0.0",
   "private": true,
   "scripts": {
+    "docs:build": "storybook build --out public/storybook && typedoc --out public/docs",
     "analyze": "ANALYZE=true next build",
     "build": "next build",
     "start": "next start -p 80",
diff --git a/tsconfig.json b/tsconfig.json
index 4750c687f..610660b5f 100644
--- a/tsconfig.json
+++ b/tsconfig.json
@@ -46,6 +46,8 @@
     ".next/types/**/*.ts"
   ],
   "exclude": [
-    "node_modules"
+    "node_modules",
+    "public/storybook",
+    "public/docs"
   ]
 }

```

File: src/core/caching/futures.ts (hunk 1 above)
Comment (henrycatalinismith) [reply src/core/caching/futures.ts:18]: Yeah not been very impressed with the reliability of typescript-plugin-mermaid tbh

**Comment by richardolsson** [comment]
I'm eager to hear your thoughts on my idea of deploying to a single app (for the dev server and preview builds on Vercel). What do you think?

**Comment by henrycatalinismith** [comment]
Christ I left this to rot a while huh. I love the idea with the single deploy. What are the steps required for making this mergeable? Here's my first draft attempt at a to-do list for myself.

- [ ] Remove the unreliable mermaid plugin and associated diagram.
- [ ] Remove the incorrect timezone-related statements.

How's that?

**Comment by richardolsson** [comment]
Love that you're back! It's been a while since I looked at this now, but I'm going to trust my previous self and assume that my last review is correct, that there are only those small changes required.

The single-deploy would be nice to get in as well, but I guess that could be a separate PR!

**Review by richardolsson — APPROVED** [review]
Lets get this merged! I can't find anything wrong with it and the documentation is a significant improvement.

Let's still try to make a single-deploy solution which hosts typedocs, storybook and NEXT.js alongside each other, which will make deployment of the preview builds faster and cheaper.

**Pull request merged**